include/swift/SIL/SILType.h

```diff
@@ -586,7 +592,42 @@ class SILType {
 
   /// Returns true if this is the AnyObject SILType;
   bool isAnyObject() const { return getASTType()->isAnyObject(); }
-  
+
+  /// Returns true if this SILType is a move only wrapper type.
+  ///
+  /// Canonical way to check if a SILType is move only. Using is/getAs/castTo
+  /// will look through moveonly-ness.
+  bool isMoveOnly() const { return getRawASTType()->is<SILMoveOnlyType>(); }
+
+  /// Return *this if already move only... otherwise, wrap the current type
+  /// within a move only type wrapper and return that. Idempotent!
+  SILType asMoveOnly() const {
+    if (isMoveOnly())
+      return *this;
+    auto newType = SILMoveOnlyType::get(getRawASTType());
+    return SILType::getPrimitiveType(newType, getCategory());
+  }
+
+  /// Return this SILType, removing moveonly-ness.
+  ///
+  /// Is idempotent.
+  SILType withoutMoveOnly() const {
+    if (!isMoveOnly())
+      return *this;
+    auto moveOnly = getRawASTType()->castTo<SILMoveOnlyType>();
+    return SILType::getPrimitiveType(moveOnly->getInnerType(), getCategory());
+  }
+
+  /// If \p otherType is move only, return this type that is move only as
+  /// well. Otherwise, returns self. Useful for propagating "move only"-ness
+  /// from a parent type to a subtype.
+  SILType copyMoveOnly(SILType otherType) const {
+    if (otherType.isMoveOnly()) {
+      return asMoveOnly();
+    }
+    return *this;
+  }
+
```

Comment:
This should never be necessary. If an aggregate type is move-only, that does not mean that projections out of it are move-only. I would remove `copyMoveOnly` and all uses of it for now.